Factories.web api

HoroshoNaVolge opened the pull request.

HoroshoNaVolge committed:
Create DAL

HoroshoNaVolge committed:
replace DAL to src

HoroshoNaVolge committed:
Add BLL Modify DAL

HoroshoNaVolge committed:
Update method correction

HoroshoNaVolge committed:
need help

Не могу зарегистриовать Hosted Service,

HoroshoNaVolge committed:
Lesson 4 finished.

Worker service added. Logger - Serilog

navferty commented on src/Factories.WebApi.DAL/Repositories/EFUnitOfWork.cs:
```diff
@@ -0,0 +1,68 @@
+        {
+            get
+            {
+                if (factoriesRepository != null)
```
non thead safe

navferty commented on src/Factories.WebApi.BLL/Controllers/UnitController.cs:
```diff
@@ -0,0 +1,72 @@
+            if (unit == null)
+                return NotFound();
+
+            return Ok(unit);
```
я бы рекомендовал разделить энтити модели, которые используются в слое ДАЛ, и те модели, которые возвращаются в публичном АПИ

HoroshoNaVolge committed:
add mapper, lock repoFactories

HoroshoNaVolge committed:
add navigation pops eager loading

HoroshoNaVolge committed:
UoF delete, Auth start writing

HoroshoNaVolge committed:
register user add

HoroshoNaVolge committed:
have problem with jwt

HoroshoNaVolge committed:
corrected Migration

HoroshoNaVolge committed:
auth corrected

HoroshoNaVolge committed:
add authentication and authorization

HoroshoNaVolge committed:
fix roles and claims + some fixes of  repos

navferty commented on src/Factories.WebApi.DAL/Repositories/UnitRepository.cs:
```diff
@@ -5,11 +5,11 @@
         private readonly FacilitiesDbContext db = db;
         private bool disposed = false;
-        public void Create(Unit item) => db.Units.Add(item);
+        public async Task CreateAsync(Unit item) => await db.Units.AddAsync(item);
```
> This method is async only to allow special value generators, such as the one used by 'Microsoft.EntityFrameworkCore.Metadata.SqlServerValueGenerationStrategy.SequenceHiLo', to access the database asynchronously. For all other cases the non async method should be used.

[Документация к методу `DbContext.AddAsync`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.addasync?view=efcore-8.0#:~:text=a%20single%20entity.-,This%20method%20is%20async%20only%20to%20allow%20special%20value%20generators%2C%20such%20as%20the%20one%20used%20by%20%27Microsoft.EntityFrameworkCore.Metadata.SqlServerValueGenerationStrategy.SequenceHiLo%27%2C%20to%20access%20the%20database%20asynchronously.%20For%20all%20other%20cases%20the%20non%20async%20method%20should%20be%20used.,-Entity%20Framework%20Core).

navferty commented on src/Factories.WebApi.DAL/Repositories/UnitRepository.cs:
```diff
@@ -33,7 +33,7 @@ public void Update(int id, Unit unit)
-        public async void Save() => await db.SaveChangesAsync();
+        public async Task SaveAsync() => await db.SaveChangesAsync();
 
         public void Dispose()
```
Не стоит диспозить экземпляр `DbContext`, так как он не был создан внутри репозитория, а получен из DI - он и будет отвечать за жизненный цикл зависимостей и вызов Dispose в нужном порядке.

navferty commented on src/Factories.WebApi.BLL/Services/WorkerService.cs:
```diff
@@ -20,7 +20,7 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 
                 await UpdateAllVolumesRandomlyAsync(tankRepository, stoppingToken);
-                tankRepository.Save();
+                tankRepository.SaveAsync();
```
Потенциальный баг - вызов асинхронного метода без `await`. Скорее всего, тут предполагалось дождаться выполнения метода. В тех же ситуациях, когда ожидание не предполагается, лучше это явно обозначить в коде, например так: `_ = DoSomethingAsync()`, давая понять что таска, которую вернул метод, не требуется для продолжения исполнения метода.

navferty commented on src/Factories.WebApi.BLL/Program.cs:
```diff
@@ -21,18 +21,18 @@ public static void Main(string[] args)
-                var issuer = configuration["Jwt:Issuer"];
-                var audience = configuration["Jwt:Audience"];
-                var key = configuration["Jwt:SecretKey"];
+                var issuer = configuration.GetSection("Jwt:Issuer").Value;
```
можно использовать класс, примерно так же как с `IOptions<T>` (с отличием в том, что `Get<T>()` может вернуть `null`):

```c#
var jwtConfig = Configuration.GetSection(JwtConfig.SectionName).Get<JwtConfig>()
  ?? throw new InvalidOperationException($"Missing required {JwtConfig.SectionName} config section!");

var issuer = jwtConfig.Issuer;


internal sealed class JwtConfig
{
  public const string SectionName = "Jwt";

  // props binded from config
}
```

HoroshoNaVolge committed:
nav props corrected, autoincr id for mapping added

plus comments resolved

HoroshoNaVolge committed:
Revert "nav props corrected, autoincr id for mapping added"

This reverts commit 3832cac97dea2d50f9a72afc082e41abfbc36e8c.

HoroshoNaVolge committed:
Revert "Revert "nav props corrected, autoincr id for mapping added""

This reverts commit e67607d419f112b932d5e78fc194d5d77a58fc6d.

HoroshoNaVolge committed:
workerService unit tests add

navferty commented on src/Factories.WebApi.BLL/Services/JWTService.cs:
```diff
@@ -7,21 +7,19 @@
 namespace Factories.WebApi.BLL.Services
 {
-    public class JwtService(string issuer, string audience, string key)
+    public class JwtService(JwtConfig jwtConfig)
```
Можно использовать [Options pattern](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-8.0#the-options-pattern): зарегистрировать опции через `builder.Services.Configure<JwtConfig>(jwtSection)`, и резолвить через DI `IOptions<JwtConfig>`.

В отличие от приведённого по ссылке примера, я бы предложил сохранять в приватное поле не `Value`, а сам объект `IOptions`: в дальнейшем это может упростить переход на `IOptionsMonitor` или `IOptionsSnapshot` (см. https://stackoverflow.com/a/61929399/16412442 )

navferty commented on src/Factories.WebApi.Tests/Factories.WebApi.Tests.csproj:
```diff
@@ -0,0 +1,29 @@
+    <PackageReference Include="coverlet.collector" Version="6.0.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
+    <PackageReference Include="Moq" Version="4.20.70" />
+    <PackageReference Include="NUnit" Version="3.14.0" />
```
NUnit продолжает достаточно широко использоваться, но в последние годы более широко распространён ~стилно-модно-молодёжный~ xUnit, рекомендую попробовать (впрочем, в целом различия не очень существенны).

navferty commented on src/Factories.WebApi.Tests/WorkerServiceTests.cs:
```diff
@@ -0,0 +1,131 @@
+            var workerService = new WorkerService(null);
+
+            // Act
+            await workerService.UpdateAllVolumesRandomlyAsync(tankRepositoryMock.Object, CancellationToken.None);
```
В случае, если в сервисе есть ошибка, потенциально ведущая к невалидному состоянию (например текущий объем меньше нуля), данная реализация теста приведёт к тому, что он будет _иногда_ падать - как повезёт с ГСПЧ.

Я бы предложил вместо `new Random` оформить рандом как внешнюю зависимость (под простым интерфейсом, вроде `interface IRandom { int NextDouble(); }`), и замокать эту зависимость. Тогда можно будет предусмотреть несколько тест-кейсов на пограничные случаи (например когда `NextDouble` возвращает минимально или максимально возможное значение).

navferty commented on src/Factories.WebApi.Tests/WorkerServiceTests.cs:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    [TestFixture]
+    public class TankVolumeUpdateTests
```
Не понятна суть этих тестов. Обычно тесты пишутся для проверки какого-то модуля (метод, класс, множество классов) кода приложения. Нет необходимости писать тесты, которые проверяют корректность синтетического кода, содержащегося внутри самих тестов (разве что за исключением локальных экспериментов при разработке, когда нужно проверить какую-то гипотезу, и проще всего запустить код именно из юнит-тестов).

HoroshoNaVolge committed:
Update authentication logic & models. Add Unit Tests for UserController

Add Interface IJwtService (first of all, to enable mocking).
Add Response models to use in UserController.
Add 401 and 404 Status Codes for failed login (wrong password, wrong username).
Add Unit Test for Login method in UserConroller.

HoroshoNaVolge committed:
comments resolved. Remade WorkerSeviceTests

navferty commented on src/Factories.WebApi.BLL/Program.cs:
```diff
@@ -29,13 +28,11 @@ public static void Main(string[] args)
 
             builder.Services.AddScoped<RoleManager<IdentityRole>>();
 
+            // ������ ��� ��� ��������� JwtConfig ��������� ��������� ������ ��� ������
```
Рекомендую всегда перед проверкой на ревью пробежаться глазами по пулл-реквесту в браузере. Во-первых, есть шанс найти какой-то лишний или недоделанный код, который был нечаянно закоммичен (например, закоментированный кусок, или какие-то временные переменные для отладки). Во-вторых, убедиться что код закоммитился корректно, например нет проблем с кодировками.

navferty commented on src/Factories.WebApi.BLL/Program.cs:
```diff
@@ -29,13 +28,11 @@ public static void Main(string[] args)
 
+            // ������ ��� ��� ��������� JwtConfig ��������� ��������� ������ ��� ������
+            var checkForValidOptions = builder.Configuration.GetSection(JwtConfig.SectionName).Get<JwtConfig>() ?? throw new InvalidConfigurationException(nameof(JwtConfig));
+            builder.Services.Configure<JwtConfig>(builder.Configuration.GetSection(JwtConfig.SectionName));
```
Я бы предложил вынести `builder.Configuration.GetSection(JwtConfig.SectionName)` в отдельную переменную, чтобы не обращаться к секции дважды.

navferty commented on src/Factories.WebApi.BLL/Services/WorkerService.cs:
```diff
@@ -34,7 +34,7 @@ public async Task UpdateAllVolumesRandomlyAsync(IRepository<Tank> tanksRepositor
             foreach (var tank in tanks)
```
`var random = new Random();` более не используется, можно удалить

navferty commented on src/Factories.WebApi.Tests/Factories.WebApi.Tests.csproj:
```diff
@@ -16,6 +16,11 @@
     <PackageReference Include="NUnit" Version="3.14.0" />
     <PackageReference Include="NUnit.Analyzers" Version="3.9.0" />
     <PackageReference Include="NUnit3TestAdapter" Version="4.5.0" />
+    <PackageReference Include="xunit" Version="2.7.0" />
```
Я бы предложил не смешивать в одной сборке с тестами NUnit и xunit. Если есть желаение использовать оба варианта в солюшне - предлагаю добавить отдельную сборку с тестами на основе xunit

HoroshoNaVolge committed:
Add SeedData, comments solved

plus some small restructurization

HoroshoNaVolge committed:
Update Program.cs

navferty commented on src/Factories.WebApi.BLL/Program.cs:
```diff
@@ -145,5 +152,19 @@ public static void Main(string[] args)
+            var roleManager = services.GetRequiredService<RoleManager<IdentityRole>>();
+            var options = services.GetRequiredService<IOptions<SeedDataOptions>>();
+
+            await dbContext.Database.MigrateAsync();
```
> While productive for local development and testing of migrations, this approach is inappropriate for managing production databases

https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/applying?tabs=dotnet-core-cli#apply-migrations-at-runtime

Я бы рекомендовал с осторожностью относиться к запуску кода миграций из самого приложения в рантайме, лучше их применять явно, с помощью командной строки, SQL скрипта или бандла (см. статью по ссылке).

Также вполне можно использовать [механизм seed data в самом EF Core](https://learn.microsoft.com/en-us/ef/core/modeling/data-seeding):
```c#
modelBuilder.Entity<Blog>().HasData(items);
```

navferty commented on src/Factories.WebApi.BLL/appsettings.json:
```diff
@@ -11,10 +11,14 @@
     "Audience": "https://localhost:7028/",
     "TokenExpiryInMinutes": 120
+  },
+  // Оставляю тут для наглядности, а так в appsettings.Development.Json (но я не уверен, если честно. Может надо добавлять в файл миграции в хешированном виде пароль?).
```
Тут могут быть разные подходы. Если оставить в `IEntityTypeConfiguration`'е конфигурацию для seed data с исходным паролем и хэшировать в рантайме при генерации очередных миграций (при выполнении команды `dotnet-ef migrations add`) - эти хэши будут отличаться каждый раз, что будет добавлять update скрипты на этого юзера. Поэтому я пришёл к тому, чтобы явно в seed data для ef core прописать сам хэш, с комментарием какому паролю он соответствует (обычно это что-то вроде user "admin" / password "admin").

navferty commented on src/Factories.WebApi.Tests/WorkerServiceNUnitTests.cs:
```diff
@@ -10,6 +10,7 @@ public class WorkerServiceNUnitTests
     {
         private Mock<IRandomService> mockRandomService;
         private Mock<IRepository<Tank>> tankRepositoryMock;
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Structure", "NUnit1032:An IDisposable field/property should be Disposed in a TearDown method", Justification = "<Ожидание>")]
```
но ведь воркер сервис реализует `IDisposable`. хотя фактически он лишь стреляет токеном отмены в методе `Dispose`, возможно, для стабильности тестов стоит обеспечить его вызов в конце выполнения тестов?

navferty commented on src/Factories.WebApi.BLL/Program.cs:
```diff
@@ -28,9 +29,15 @@ public static void Main(string[] args)
+            var jwtOptionsSectionName = builder.Configuration.GetSection(JwtOptions.SectionName);
+            var seedDataOptionsSectionName = builder.Configuration.GetSection(SeedDataOptions.SectionName);
+
+            var checkForValidJwtOptions = jwtOptionsSectionName.Get<JwtOptions>() ?? throw new InvalidOperationException(nameof(JwtOptions));
```
Возможно, стоит уточнить текст исключения (например что-то вроде "Config section {SectionName} must be set up")

navferty commented on src/Factories.WebApi.BLL/Program.cs:
```diff
@@ -28,9 +29,15 @@ public static void Main(string[] args)
-            // ������ ��� ��� ��������� JwtConfig ��������� ��������� ������ ��� ������
-            var checkForValidOptions = builder.Configuration.GetSection(JwtConfig.SectionName).Get<JwtConfig>() ?? throw new InvalidConfigurationException(nameof(JwtConfig));
-            builder.Services.Configure<JwtConfig>(builder.Configuration.GetSection(JwtConfig.SectionName));
+            var jwtOptionsSectionName = builder.Configuration.GetSection(JwtOptions.SectionName);
```
Переменная называется ...SectionName, но ведь это не имя секции, а сама секция. Возможно, опечатка?

HoroshoNaVolge committed:
comments solved, add seed data for FacilitiesDbContext

navferty commented on src/Factories.WebApi.BLL/Database/UsersDbContext.cs:
```diff
@@ -1,9 +1,19 @@
+
+            // Я для примера добавлю сюда seed ролей в таблицу ролей, но самого админа в OnModelCreating корректно не добавить,
+            // т.к. нужно ему добавить роль и клеймы асинхронно через UserManager (в OnModelCreating не сделать асинхронно).
+            modelBuilder.Entity<IdentityRole>().HasData(new IdentityRole { Id = Guid.NewGuid().ToString(), Name = "Admin", NormalizedName = "ADMIN" },
```
`Guid.NewGuid().ToString()` вызывается заново каждый раз при построении моделе. В итоге когда ты будешь добавлять следующую миграцию, наряду с нужными изменениями он также добавит update (или remove+add).

navferty commented on src/Factories.WebApi.BLL/Database/UsersDbContext.cs:
```diff
@@ -1,9 +1,19 @@
+        {
+            base.OnModelCreating(modelBuilder);
+
+            // Я для примера добавлю сюда seed ролей в таблицу ролей, но самого админа в OnModelCreating корректно не добавить,
```
Но почему собственно не добавить юзера тут же, через `Entity<User>().HasData(new { ... })` например. За образец можно взять состояние БД после добавления через UserManager (или посмотреть [сорцы](https://github.com/dotnet/aspnetcore/blob/main/src/Identity/Extensions.Core/src/UserManager.cs)).

HoroshoNaVolge committed:
comments solved

HoroshoNaVolge committed:
small fix

navferty commented on src/Factories.WebApi.BLL/Database/UsersDbContext.cs:
```diff
@@ -10,10 +10,42 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
-            // т.к. нужно ему добавить роль и клеймы асинхронно через UserManager (в OnModelCreating не сделать асинхронно).
-            modelBuilder.Entity<IdentityRole>().HasData(new IdentityRole { Id = Guid.NewGuid().ToString(), Name = "Admin", NormalizedName = "ADMIN" },
-                                                         new IdentityRole { Id = Guid.NewGuid().ToString(), Name = "User", NormalizedName = "USER" }); 
+            modelBuilder.Entity<IdentityRole>().HasData(new IdentityRole { Id = "70a2bf46-7e0a-4559-a625-5b3ca954b3cf", Name = "Admin", NormalizedName = "ADMIN" },
```
В целом, подход ок, но я бы предложил вынести конфигурации разных сущностей в отдельные классы, реализующие `IEntityTypeConfiguration<TEntity>`, см. [статью](https://learn.microsoft.com/en-us/ef/core/modeling/#grouping-configuration).

HoroshoNaVolge committed:
comment solved

navferty commented on src/Factories.WebApi.BLL/Migrations/UsersDbContextModelSnapshot.cs:
```diff
@@ -155,15 +155,15 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                             Id = "fe342990-c53a-4bb9-89b6-4b4482e956fb",
                             AccessFailedCount = 0,
-                            ConcurrencyStamp = "0bbf3574-511d-46c5-8aba-5d0d2735b11b",
+                            ConcurrencyStamp = "2b5ddc62-ae7e-4ffb-8d1f-9053ebc7ceea",
```
Попробуй добавить новую миграцию. В идеале она должна сгенерироваться пустой. Если в той миграции будет update скрипт для `ConcurrencyStamp` и `SecurityStamp`, возможно стоит эти поля тоже явно прописать в конфигурации для seed data.

HoroshoNaVolge committed:
Add Dapper for TanksRepository

plus correct seed data for usersdb

HoroshoNaVolge committed:
Update WorkerService.cs

HoroshoNaVolge committed:
Mapperly instead of AutoMapper

Plus added dapper reso for other entitites

navferty commented on src/Factories.WebApi.BLL/Program.cs:
```diff
@@ -76,11 +76,26 @@ public static async Task Main(string[] args)
-                             .AddScoped<IRepository<Factory>, FactoryRepository>();
-
-            builder.Services.AddAutoMapper(typeof(MappingProfile));
+            // Реализация на EF Core
```
Можно добавить флаг в appsettings.json, и выбирать реализацию в зависимости от конфигурации.

navferty commented on src/Factories.WebApi.DAL/Repositories/DapperRepositories/FactoryRepositoryDapper.cs:
```diff
@@ -0,0 +1,62 @@
+
+        public async Task SaveAsync()
+        {
+            await Task.Delay(0);
```
Можно использовать `await Task.CompletedTask;`.

Есть также вариант `await Task.Yield();`, если нужно, чтобы продолжение метода точно выполнялось асинхронно.

navferty commented on src/Factories.WebApi.DAL/Repositories/DapperRepositories/FactoryRepositoryDapper.cs:
```diff
@@ -0,0 +1,62 @@
+        public async Task SaveAsync()
+        {
+            await Task.Delay(0);
+            // Насколько я понял, транзакции не нужны для простых операций СRUD, а Dapper сам сохраняет изменения в БД при запросах.
```
Да, возможно имеет смысл перенести вызов SaveChanges в методы, которые собственно меняют данные. Иначе, например, при таком коде

```
var someNewItem = new Item
{
  Id = 42,
  /* other props initialization */
};
repo.Add(someNewItem);

// found будет null, если не было вызова SaveChanges
var found = repo.Find(42);
```

HoroshoNaVolge committed:
comments solved, db names to lower, migrations remake

HoroshoNaVolge committed:
add docker containerization

navferty commented on docker-compose.yml:
```diff
@@ -0,0 +1,25 @@
+    image: postgres:13
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: corsar
```
предлагаю перенести в .env файл (который, в свою очередь, добавить в гитигнор)

navferty commented on src/Factories.WebApi.BLL/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+WORKDIR /src
+COPY ["src/Factories.WebApi.BLL/Factories.WebApi.BLL.csproj", "src/Factories.WebApi.BLL/"]
+COPY ["src/Factories.WebApi.DAL/Factories.WebApi.DAL.csproj", "src/Factories.WebApi.DAL/"]
+RUN dotnet restore "./src/Factories.WebApi.BLL/Factories.WebApi.BLL.csproj"
```
плюсик за отдельное копирование csproj и нугет рестор, для ускорения билда при кэшировании этих слоёв. рекомендую также почитать [статью](https://andrewlock.net/optimising-asp-net-core-apps-in-docker-avoiding-manually-copying-csproj-files-part-2/#the-new-improved-solution), я в своём проекте использовал немного более универсальные команды, чтобы не добавлять все csproj по отдельности:

```
# Copy only csproj files first before nuget restore to force caching nuget restore layers
# https://andrewlock.net/optimising-asp-net-core-apps-in-docker-avoiding-manually-copying-csproj-files-part-2/#the-new-improved-solution

COPY src/*/*.csproj ./
RUN for file in $(ls *.csproj); do mkdir -p src/${file%.*}/ && mv $file src/${file%.*}/; done

COPY tests/*/*.csproj ./
RUN for file in $(ls *.csproj); do mkdir -p tests/${file%.*}/ && mv $file tests/${file%.*}/; done

# Restore NuGet packages for entire solution, so that this layer could be reused by any of projects from this solution
RUN dotnet restore

# Copy the source
COPY src/ ./src/

# Build the application
RUN dotnet build src/WorkerService/WorkerService.csproj -c Release --no-restore
```

navferty commented on src/Factories.WebApi.BLL/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+ARG BUILD_CONFIGURATION=Release
+RUN dotnet publish "./Factories.WebApi.BLL.csproj" -c $BUILD_CONFIGURATION -o /app/publish /p:UseAppHost=false
+
+FROM base AS final
```
Я бы рекомендовал, если возможно, запускать приложение не от root'а, что-то вроде такого:

```
RUN addgroup --gid 2000 --group "group1"
RUN adduser  --gid 2000 --uid 1000 "user1" 

RUN chown user1:group1  /app
USER user1:group1
```

[Статья](https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b) по теме.

navferty commented on src/Factories.WebApi.BLL/appsettings.json:
```diff
@@ -7,13 +7,14 @@
   "ConnectionStrings": {
-    "DefaultConnection": "Host=localhost;Port=5432;Database=facilitiesDb;Username=postgres;Password=corsar",
-    "UsersConnection": "Host=localhost;Port=5432;Database=usersDb;Username=postgres;Password=corsar"
+    "DefaultConnection": "Host=postgres-container;Port=5432;Database=postgres;Username=postgres;Password=corsar",
```
дать возможность переопределять конфиги через переменные среды. Желательно сразу использовать префикс, во избежание коллизий, например `DOTNETLEARNING_` (переменная будет вида `DOTNETLEARNING_ConnectionStrings__DefaultConnection=Host=postgres-container;Port=5432;<...>`

HoroshoNaVolge committed:
add .env

HoroshoNaVolge committed:
appsettings and docker-compose correction

HoroshoNaVolge committed:
Update README.md

HoroshoNaVolge committed:
add redis cache to dapper repos

HoroshoNaVolge committed:
Merge branch 'Factories.WebApi' of https://github.com/HoroshoNaVolge/dotnetLearning into Factories.WebApi

HoroshoNaVolge committed:
RabbitMQ add for BackgroundService

navferty commented on src/RunInParallelJsonPlaceholder/Program.cs:
```diff
@@ -0,0 +1,110 @@
+                    return await taskFunc();
+                }
+
+                catch (HttpRequestException ex)
```
метод расширения предполагается быть универсальным, не должен знать что используется именно для запросов http

navferty commented on src/RunInParallelJsonPlaceholder/Program.cs:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    public static class ParallelExtensions
```
Предлагаю реализовать этот класс в отдельном проекте, и задокументировать его (как если его предполагается публиковать в виде nuget-пакета для общего использования)

navferty commented on src/RunInParallelJsonPlaceholder/Program.cs:
```diff
@@ -0,0 +1,110 @@
+    public static class ParallelExtensions
+    {
+        public static async Task<IReadOnlyCollection<T>> RunInParallel<T>(
+            this IEnumerable<Func<Task<T>>> tasks, int maxParallelTasks = 4)
```
Предлагаю также добавить опциональный параметр (например enum), который позволял бы настраивать стратегию обработки исключений:

* немедленно пробрасывать исключение в вызывающий метод, при этом желательно продумать отмену других выполняющихся задач (и конечно отмену тех, которые ещё даже не успели стартовать)
* складывать все исключения, и пробрасывать их в вызывающий код как `AggregatedException`
* игнорировать исключения, и возвращать коллекцию результатов только тех задач, которые успешно завершились
* возможно, сделать ещё один метод-расширения, с другим типом возвращаемого значения, который позволит вызывающему коду как получить результаты успешно выполненных задач, так и иметь возможность увидеть ошибки

Возможно, также стоит добавить опциональный аргумент типа `ILogger`, так как не во всех приложениях есть возможность читать консоль (например, ASP.NET Web API, либо десктопное приложение WinForms/WPF).

navferty commented on src/RunInParallelJsonPlaceholder/Program.cs:
```diff
@@ -0,0 +1,110 @@
+        public static async Task<IReadOnlyCollection<T>> RunInParallel<T>(
+            this IEnumerable<Func<Task<T>>> tasks, int maxParallelTasks = 4)
+        {
+            var taskLists = tasks.ToList();
```
Зачем вызывать `.ToList()` перед тем, как выполнить цикл `foreach` по этой же коллекции?

navferty commented on src/RunInParallelJsonPlaceholder/Program.cs:
```diff
@@ -0,0 +1,110 @@
+
+                catch (HttpRequestException ex)
+                {
+                    await Console.Out.WriteLineAsync($"Network error: {ex.Message}");
```
см. комментарий выше про логирование - не во всех приложениях есть возможность читать вывод в консоль

navferty commented on src/RunInParallelJsonPlaceholder/Program.cs:
```diff
@@ -0,0 +1,110 @@
+            {
+                var task = FetchPostAsync(postId++);
+                tasks.Add(() => task);
+                var result = await task;
```
Если в каждой итерации цикла происходит ожидание того, когда загрузится очередной пост, в чём заключается "параллелизм" метода `LoadPostsInParallel`?

HoroshoNaVolge committed:
fixed docker compose, GetAsync method added

HoroshoNaVolge committed:
Update README.md

HoroshoNaVolge committed:
Update README.md

HoroshoNaVolge committed:
Update README.md
